src/main/kotlin/it/pagopa/wallet/cdc/PaymentWalletsLogEventsStream.kt

```diff
@@ -44,22 +51,49 @@ class PaymentWalletsLogEventsStream(
                                 Aggregation.project(changeStreamOptionsConfig.project)
                             )
                         )
-                        .resumeAt(Instant.now())
+                        .resumeAt(redisResumePolicyService.getResumeTimestamp())
                         .build(),
                     BsonDocument::class.java
                 )
-                .flatMap {
-                    Mono.defer {
-                            walletPaymentCDCEventDispatcherService.dispatchEvent(
-                                it.raw?.fullDocument?.toBsonDocument()
-                            )
-                        }
-                        .onErrorResume {
-                            logger.error("Error during event handling : ", it)
-                            Mono.empty<BsonDocument>()
-                        }
-                }
+                // Process the elements of the Flux
+                .flatMap { processEvent(it) }
+                // Save resume token every n emitted elements
+                .index()
+                .flatMap { saveCdcResumeToken(it) }
+                .doOnError { logger.error("Error listening to change stream: ", it) }
 
         return flux
     }
+
+    private fun processEvent(event: ChangeStreamEvent<BsonDocument>): Mono<BsonDocument> {
+        return Mono.defer {
+                walletPaymentCDCEventDispatcherService.dispatchEvent(
+                    event.raw?.fullDocument?.toBsonDocument()
+                )
+            }
+            .onErrorResume {
+                logger.error("Error during event handling : ", it)
+                Mono.empty()
+            }
+    }
+
+    private fun saveCdcResumeToken(
+        changeEventFluxIndex: Tuple2<Long, BsonDocument>
+    ): Mono<BsonDocument> {
+        return Mono.defer {
+                if (changeEventFluxIndex.t1.absoluteValue.plus(1).mod(saveInterval) == 0) {
```

Comment:
```suggestion
                 if ((idx + 1 ) % (saveInterval) == 0L) {
```
this can be simplified avoiding method invocation and making formula more readable.
Moreover [index](https://projectreactor.io/docs/core/release/api/reactor/core/publisher/Flux.html#index-java.util.function.BiFunction-) methods returns a 0-based incrementing long so no need to calculate absolute value

Reply:
Absolute value has been introduced to handle overflow scenario ( I know it is unlikely to happen :)), we could get rid of it. For the readability of the formula, I think that the actual approach is more readable

Reply:
index is a long value, I think overflow can safely be ignored here :D 

For the readability IDK, I personally find read method concatenation less immediate that the math formula version

Reply:
Absolute value removed in [b65daa4](https://github.com/pagopa/pagopa-payment-wallet-cdc-service/pull/11/commits/b65daa4a724803121f3aaad9f15081ce9ba4f1cb)